cmd/snap-device-helper/snap-device-helper.c

```diff
@@ -175,11 +176,26 @@ int snap_device_helper_run(const struct sdh_invocation *inv) {
     char fullsubsystem[PATH_MAX] = {0};
     sc_must_snprintf(sysdevsubsystem, sizeof(sysdevsubsystem), "%s/sys/%s/subsystem", sysroot, devpath);
     if (readlink(sysdevsubsystem, fullsubsystem, sizeof(fullsubsystem)) < 0) {
-        die("cannot read symlink %s", sysdevsubsystem);
-    }
-    char *subsystem = basename(fullsubsystem);
-    if (sc_streq(subsystem, "block")) {
-        devtype = S_IFBLK;
+        if (errno == ENOENT && sc_streq(action, "remove")) {
+            // on removal the devices are going away, so it is possible that the
+            // symlink is already gone, in which case try guessing the type like
+            // the old shell-based snap-device-helper did:
+            //
+            // > char devices are .../nvme/nvme* but block devices are
+            // > .../nvme/nvme*/nvme*n* and .../nvme/nvme*/nvme*n*p* so if have a
+            // > device that has nvme/nvme*/nvme*n* in it, treat it as a block
+            // > device
+            if ((fnmatch("*/block/*", devpath, 0) == 0) || (fnmatch("*/nvme/nvme*/nvme*n*", devpath, 0) == 0)) {
```

Comment:
It might be safer to use `FNM_PATHNAME` in the second invocation, but then we'd need to know the path prefix... And I don't really see a risk of that `/nvme*/` matching a deeper directory tree, so (unless you have some idea) I'd leave it as it is.

Reply:
This change matches what the old shell-script based snap-device-helper used so this should be fine.  https://github.com/snapcore/snapd/commit/db7371925d53093cbb8c492200af8112067f8fd7#diff-2694e7a2145198b15b6774227aa01c64311b459fa86059e281fcd78e17de240cL58